### PR TITLE
Add wrappers for ArborX::Sphere

### PR DESCRIPTION
**Pull request by Rombur**
ArborX supports three different geometrical objects: `Box`,  `Point`, and `Sphere`. We already have wrappers for `Box` and `Point`, this PR adds wrappers for `Sphere`.

### CODE REVIEW

File: include/deal.II/arborx/access_traits.h [hunk 1]
```diff
@@ -205,6 +207,100 @@ namespace ArborXWrappers
   private:
     unsigned int n_nearest_neighbors;
   };
+
+
+
+  /**
+   * Base class for Sphere-based predicates providing basic functionality for
+   * derived classes, not supposed to be used on its own.
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:215]:
```suggestion
   * derived classes; not supposed to be used on its own.
```

File: include/deal.II/arborx/access_traits.h [hunk 2]
```diff
@@ -205,6 +207,100 @@ namespace ArborXWrappers
...
+   */
+  class SpherePredicate
+  {
+  protected:
+    /**
+     * Constructor. @p spheres is a list of spheres, i.e. center and radius
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:221]:
```suggestion
     * Constructor. @p spheres is a list of spheres, i.e. a center and radius
```

File: include/deal.II/arborx/access_traits.h [hunk 3]
```diff
@@ -205,6 +207,100 @@ namespace ArborXWrappers
...
+     * pair, used by the predicate.
+     */
+    template <int dim, typename Number>
+    SpherePredicate(
+      const std::vector<std::pair<dealii::Point<dim, Number>, Number>>
+        &spheres);
+
+    /**
+     * Number of spheres stored in the structure.
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:230]:
```suggestion
     * The number of spheres stored in the structure.
```

File: include/deal.II/arborx/access_traits.h [hunk 4]
```diff
@@ -205,6 +207,100 @@ namespace ArborXWrappers
...
+    // We need these since we inherit privately to avoid polymorphic use.
+    using SpherePredicate::get;
+    using SpherePredicate::size;
+  };
+
+
+
+  /**
+   * This class defines a predicate used by ArborXWrappers::BVH to determine for
+   * given spheres which are the nearest bounding boxes/points among the ones
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:274]:
```suggestion
   * This class defines a predicate used by ArborXWrappers::BVH to determine, for the
   * given spheres, which are the nearest bounding boxes/points among the ones
```

File: include/deal.II/arborx/access_traits.h [hunk 5]
```diff
@@ -370,6 +497,59 @@ namespace ArborX
...
+   * This struct allows ArborX to use SphereIntersectPredicate as a predicate.
+   */
+  template <>
+  struct AccessTraits<dealii::ArborXWrappers::SphereIntersectPredicate,
+                      PredicatesTag>
+  {
+    using memory_space = Kokkos::HostSpace;
+
+    /**
+     * Number of Point stored in @p sph_intersect.
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:512]:
```suggestion
     * The number of points stored in @p sph_intersect.
```

File: include/deal.II/arborx/access_traits.h [hunk 6]
```diff
@@ -370,6 +497,59 @@ namespace ArborX
...
+   * This struct allows ArborX to use SphereNearestPredicate as a predicate.
+   */
+  template <>
+  struct AccessTraits<dealii::ArborXWrappers::SphereNearestPredicate,
+                      PredicatesTag>
+  {
+    using memory_space = Kokkos::HostSpace;
+
+    /**
+     * Number of spheres stored in @p sph_nearest.
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:538]:
```suggestion
     * The number of spheres stored in @p sph_nearest.
```

File: source/arborx/access_traits.cc [hunk 7]
```diff
@@ -146,6 +146,69 @@ namespace ArborXWrappers
...
+    const std::vector<std::pair<dealii::Point<dim, Number>, Number>>
+      &dim_spheres)
+  {
+    static_assert(dim != 1, "dim equal to one is not supported.");
+
+    const unsigned int size = dim_spheres.size();
+    spheres.reserve(size);
+    for (unsigned int i = 0; i < size; ++i)
+      {
+        spheres.emplace_back(std::make_pair(
```
Comment (jppelteret) [inline source/arborx/access_traits.cc:162]: Do you wish to add here the same comment that you make later, i.e.
```
    // ArborX assumes that the center coordinates and the radius use float and
    // the sphere is 3D
```
?

File: source/arborx/access_traits.cc [hunk 8]
```diff
@@ -146,6 +146,69 @@ namespace ArborXWrappers
   {
     return n_nearest_neighbors;
   }
+
+  // ------------------- SpherePredicate ------------------- //
+  template <int dim, typename Number>
+  SpherePredicate::SpherePredicate(
+    const std::vector<std::pair<dealii::Point<dim, Number>, Number>>
+      &dim_spheres)
```
Comment (jppelteret) [inline source/arborx/access_traits.cc:154]: Is there a particular reason why you use the name `dim_spheres` here but just `spheres` elsewhere? Just curious...

**Review by jppelteret — APPROVED** [review]
Just a few very superficial comments from me, otherwise I think this looks good. Thanks @Rombur!

**Comment by Rombur** [comment]
I have updated the documentation

File: source/arborx/access_traits.cc (hunk 8 above)
Comment (jppelteret) [reply source/arborx/access_traits.cc:154]: Ah, I see. Thanks a lot for the explanation!

**Pull request merged**